### PR TITLE
Create one script to setup and run the cerebro2 server

### PR DESCRIPTION
**Pull request by passiweinberger**
For easier use....

### CODE REVIEW

**Comment by rhyolight** [comment]
@chetan51 Please review.

File: run_server.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env python
+
+import os
+import sys
+import webbrowser
+import subprocess
+"""
+## Usage
+Make sure the 'nupic.cerebro2.server directory' is extracted in here and all requirements are installed. 
+(Furter information : see the READMEs)
```
Comment (rhyolight) [inline run_server.py]: typo `Furter`

**Comment by rhyolight** [comment]
@passiweinberger Please sign our [contributor license](http://numenta.org/licenses/cl/). Also, I'll make some notes and point out examples, but you'll need to make some naming changes to adhere to our [Python coding standards](https://github.com/numenta/nupic/wiki/Python-Style-Guide#function-method-and-instance-variable-names).

File: run_server.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    		"Patcher().patchTP(tp)\n\n"
+    		
+		"This scripc also assumes that you have the 'nupic.cerebro2.server' folder inside this directory!\n"
+		"For further information read the READMEs in these two folders!\n\n"
+		"If last lines for Exit Cmd uncommented:\n"
+		"Press C-c to kill all the processes. \n"
+		"(CAUTION: executes 'killall python' !!)\n")
+
+		
+def server_setup(StartDir):
```
Comment (rhyolight) [inline run_server.py]: `startDir`

File: run_server.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    if not os.path.isdir(ServerDir):
+      print "Could not find the nupic.cerebro2.server directory! Please copy it into this Directory!\nTerminated."
+      return 1
+    else:
+      os.chdir(ServerDir)
+      print "Server is being set up...\n"
+      subprocess.Popen(["python", "server.py"])
+
+
+def server_call(StartDir, port):
```
Comment (rhyolight) [inline run_server.py]: `startDir`

File: run_server.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    StaticDir = os.path.join(StartDir, 'static')
+    os.chdir(StaticDir)
+    print "Calling the server at port number %s" % port 
+    command = "python -m SimpleHTTPServer " + str(port)    
+    subprocess.Popen(["python", "-m", "SimpleHTTPServer", port])    
+    print "Server is up and running at port %s!\n" % port
+
+
+def open_browser(port):
```
Comment (rhyolight) [inline run_server.py]: `openBrowser`

File: run_server.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    #Setting up the server in nupic.cerebro2.server module:
+    server_setup(StartDir)
+    #Calling the server for the provided port:
+    server_call(StartDir, port) 
+    #Opening the browser at the port:
+    open_browser(port)
+
+    #Exit command:
+    #if KeyboardInterrupt:
+    #	os.system("gnome-terminal -e 'killall python'")
```
Comment (rhyolight) [inline run_server.py]: Remove this commented code or find a system-independent way to kill all?

File: run_server.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+Then, run:
+
+    ./run_server <PortNumber>
+
+### Screenshots
+
+To enable taking screenshots with the `p` keyboard shortcut, open `js/classes/abstract_visualization.js` and uncomment the line under the note about screenshots.
+
+@author: passiweinberger
```
Comment (rhyolight) [inline run_server.py]: Please remove, but feel free to add an `AUTHORS.md` to the root directory if you want to claim some ownership.

Comment (passiweinberger) [inline run_server.py]: Is that appropriate ? Saw most scripts having it so i adapted this...

File: run_server.py (hunk 5 above)
Comment (passiweinberger) [inline run_server.py]: Done

**Comment by passiweinberger** [comment]
Got all changed. Btw, i was wondering whether to call it CLA or HTM now ? Think it has been changed to HTM as of the new white paper? Thanks :)

<div>-------- Original message --------</div><div>From: Matthew Taylor <notifications@github.com> </div><div>Date:03/16/2015  19:21  (GMT+01:00) </div><div>To: "numenta/nupic.cerebro2" <nupic.cerebro2@noreply.github.com> </div><div>Cc: Pascal Weinberger <passiweinberger@gmail.com> </div><div>Subject: Re: [nupic.cerebro2] Create one script to setup and run ther cerebro2
  server (#71) </div><div>
</div>@passiweinberger Please sign our contributor license. Also, I'll make some notes and point out examples, but you'll need to make some naming changes to adhere to our Python coding standards.

**Comment by rhyolight** [comment]
> Btw, i was wondering whether to call it CLA or HTM now ?

HTM please.

File: run_server.py (hunk 6 above)
Comment (rhyolight) [inline run_server.py]: I don't like having authors labelled in the source file. Because we have git, we don't need it to see who's worked on what. And people don't keep it up to date so you can't depend on it.

**Comment by passiweinberger** [comment]
Okay.. but it might be confusing as the patcher calls it CLA, so we should change it everewhere at once then right? until then I'll insert (HTM)in the description to CLA, but I don't think calling it HTM only, but then point to the patchCLA function is really clear... :s

**Comment by rhyolight** [comment]
Yes, we should change it everywhere. We're still tracking down all the "CLA" references.

**Comment by passiweinberger** [comment]
Ok then Ill change it here and the pather Fuction should be too

**Comment by rhyolight** [comment]
@passiweinberger Please leave the `patchCLAModel` function name the same for now. In the OPF we still have a class called `CLAModel` that we cannot change until we update NuPIC serialization process, and that's what the function refers to. It would be confusing to change this at this point. 

If you see references to the general term "CLA" in comments, those can be changed to "HTM". Sorry for the confusion.

**Comment by passiweinberger** [comment]
oki doki :) was just about to pull the change on the .server file as well ..

**Comment by passiweinberger** [comment]
Should be all right now :)

File: run_server.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/env python
+
+import os
+import sys
+import webbrowser
+import subprocess
+"""
+## Usage
```
Comment (chetan51) [inline run_server.py]: I recommend consolidating the information in the docstring and the `DESCRIPTION` below, to avoid repeating and having to maintain two copies of the info.

**Comment by chetan51** [comment]
This README is designed to depend on the `nupic.cerebro2.server` README, to avoid repeating information between the two (and having to maintain them separately). I think it's convenient to have a single `run_server.py` script (thanks for that!) but instead of changing this README, can we just add a note pointing out `run_server.py` as a convenient alternative?

**Comment by passiweinberger** [comment]
Sth like that ?

File: README.md [hunk 8]
```diff
@@ -4,16 +4,36 @@ A web-based visualization and debugging platform for NuPIC.
 
 ## Usage
 
-Set up [cerebro2.server](https://github.com/numenta/nupic.cerebro2.server) to export your model state.
+Set up cerebro2.server to export your model state.
```
Comment (chetan51) [inline README.md]: Unnecessary change.

File: README.md [hunk 9]
```diff
@@ -4,16 +4,36 @@ A web-based visualization and debugging platform for NuPIC.
...
 
 Then, run:
 
     cd static
     python -m SimpleHTTPServer 8000
-
```
Comment (chetan51) [inline README.md]: Unnecessary change.

**Comment by chetan51** [comment]
I was thinking, just a note in the README saying "Check out `run_server.py` for a quicker setup process." (Since `run_server.py` already has usage information in it.)

**Comment by passiweinberger** [comment]
oh okay sorry did't get that.

**Comment by passiweinberger** [comment]
Allright with that ?

File: README.md [hunk 10]
```diff
@@ -10,11 +9,13 @@ Then, run:
...
 Finally, visit the following URL in your browser:
 
     http://localhost:8000
-
+ 
 ### Screenshots
 
 To enable taking screenshots with the `p` keyboard shortcut, open `js/classes/abstract_visualization.js` and uncomment the line under the note about screenshots.
+
+### Quicker Solution can be found in run_server.py
```
Comment (chetan51) [inline README.md]: Cleaner:

    ### Quicker setup
    
    See `run_server.py` for a quicker setup.

Also please remove the other extraneous changes in this README.

Comment (passiweinberger) [inline README.md]: oki doki :) No changes left. the rest is yours copied

File: run_server.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    port = sys.argv[1]
+    # Setting up the server in nupic.cerebro2.server module:
+    serverSetup(startDir)
+    # Calling the server for the provided port:
+    serverCall(startDir, port) 
+    # Opening the browser at the port:
+    openBrowser(port)
+
+    # Exit command:
+    #if KeyboardInterrupt:
```
Comment (chetan51) [inline run_server.py]: Please uncomment or remove commented-out code.

**Comment by passiweinberger** [comment]
Okay. So we don't comment here?

**Comment by chetan51** [comment]
@passiweinberger I wasn't referring to comments, but rather _commented-out code_. Comments are fine, but we avoid leaving commented-out (dead) code in there.

**Comment by passiweinberger** [comment]
Oh well i thought this would be nice for people to have by hand if they want to stop everything by a key comb. As I ran cerebro my computer got really stuck some times an i just wanted to kill all the processes to see whats wrong. But typing and opening another terminal lagged too so this was helpful to me atleast. But having it in as-is might cause problems though so I  though this would be a easy solution ... ?

<div>-------- Original message --------</div><div>From: Chetan Surpur <notifications@github.com> </div><div>Date:03/18/2015  19:56  (GMT+01:00) </div><div>To: "numenta/nupic.cerebro2" <nupic.cerebro2@noreply.github.com> </div><div>Cc: Pascal Weinberger <passiweinberger@gmail.com> </div><div>Subject: Re: [nupic.cerebro2] Create one script to setup and run the cerebro2
  server (#71) </div><div>
</div>@passiweinberger I wasn't referring to comments, but rather commented-out code. Comments are fine, but we avoid leaving commented-out (dead) code in there.

**Comment by chetan51** [comment]
Why not have it enabled by default?

**Comment by passiweinberger** [comment]
kills ALL .py processes... one might have sth important running what should not be killed; a swarming or sth like that in the background which would terminate too. So it should be up to oneself to decide. I point to it in the description, so its easy to switch on if wnated

**Comment by passiweinberger** [comment]
oh but we could just kill only the server and client processes right ? but im not sure if they are always the same ?

**Comment by chetan51** [comment]
Okay, makes sense. Thanks!
